depth_image_proc/src/nodelets/convert_metric.cpp

```diff
@@ -1,140 +1,145 @@
+// modification, are permitted provided that the following conditions
+// are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of the Willow Garage nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include <rclcpp/rclcpp.hpp>
 #include <image_transport/image_transport.h>
-#include <sensor_msgs/image_encodings.h>
-#include <boost/thread.hpp>
+#include <sensor_msgs/image_encodings.hpp>
+#include <depth_image_proc/visibility.h>
+#include <cmath>
+#include <memory>
+#include <limits>
 
-namespace depth_image_proc {
+namespace depth_image_proc
+{
 
 namespace enc = sensor_msgs::image_encodings;
 
-class ConvertMetricNodelet : public nodelet::Nodelet
+class ConvertMetricNode : public rclcpp::Node
 {
+public:
+  DEPTH_IMAGE_PROC_PUBLIC ConvertMetricNode();
+
+private:
   // Subscriptions
-  boost::shared_ptr<image_transport::ImageTransport> it_;
+  std::shared_ptr<image_transport::ImageTransport> it_;
   image_transport::Subscriber sub_raw_;
 
   // Publications
-  boost::mutex connect_mutex_;
+  std::mutex connect_mutex_;
   image_transport::Publisher pub_depth_;
 
-  virtual void onInit();
+  void connectCb(rclcpp::Node::SharedPtr node);
 
-  void connectCb();
+  void depthCb(const sensor_msgs::msg::Image::ConstSharedPtr & raw_msg);
 
-  void depthCb(const sensor_msgs::ImageConstPtr& raw_msg);
+  rclcpp::Logger logger_ = rclcpp::get_logger("ConvertMetricNode");
 };
 
-void ConvertMetricNodelet::onInit()
+ConvertMetricNode::ConvertMetricNode()
+: Node("ConvertMetricNode")
 {
-  ros::NodeHandle& nh = getNodeHandle();
-  it_.reset(new image_transport::ImageTransport(nh));
+  rclcpp::Node::SharedPtr node = std::shared_ptr<rclcpp::Node>(this);
```

Comment:
While this appears to work for you, it looks to be unsafe based on the documentation is here: https://en.cppreference.com/w/cpp/memory/enable_shared_from_this

The docs indicate: 
> enable_shared_from_this provides the safe alternative to an expression like std::shared_ptr<T>(this), which is likely to result in this being destructed more than once by multiple owners that are unaware of each other (see example below)

The first workaround that I thought of is that we can just use `shared_from_this`, but using it in the constructor could also lead to potentially undefined behavior.

Reply:
My proposed workaround is to allow `message_filters` and `image_transport` take a raw pointer to `Node`, because it's unlikely that the `SharedPtr` is providing real benefit or safety here.